add pattern matching docs (#334)
arr-ai/arrai: docs/tutorial/binding.md

We'll learn more about `=>` in a later tutorial.

## Pattern matching

Both let-bindings and function parameters support pattern matching. This is a
very powerful mechanism to extract elements from a complex structure and also
restrict what values may be used as input.

> Try out the following examples to use pattern in array:

Reviewer: s/in array/with arrays/. Same for tuple, etc.

```arrai
@> let [] = []; 1
@> let [a, b, c] = [1, 2, 3]; b
@> let arr = [1, 2]; let [a, b] = arr; b
@> let [x, x] = [1, 1]; x
@> let [x, x] = [1, 2]; x                  # should fail
```

in tuple:
```arrai
@> let () = (); 1
@> let (a: x, b: y) = (a: 4, b: 7); x
@> let (a: x, b: x) = (a: 4, b: 4); x
@> let (:x) = (x: 1); x
```

in dictionary:
```arrai
@> let {"a": f, "b": k} = {"a": 1, "b": 2}; [f, k]
```

and in set:
```arrai
@> let {} = {}; 1
@> let {a, b, c} = {1, 2, 3}; a
@> let {a, 42} = {3, 42}; a
```

> Also, nested pattern is supported as:

Reviewer: "patterns are"

```arrai
@> let [[x, y], z] = [[1, 2], 3]; x
@> let [{"a": x}, (b: y), z] = [{"a": 1}, (b: 2), 3]; [x, y, z]
```

> Underscore `_` matches any items and ignore it.

Reviewer: s/items and ignore it/value and ignores it/

```arrai
@> let [x, _, _] = [1, 2, 3]; x
@> let [_, x, _] = [1, 2, 3]; x
```

> Name within parentheses like `(x)` refers to the value `x` which is supposed to be defined before.

Reviewer: ```suggestion
A name within parentheses like `(x)` refers to the value bound to the name `x`.
If `x` isn't bound to a value, the match fails.
```

```arrai
@> let x = 3; let [b, x] = [2, 4]; x
@> let x = 3; let [b, (x)] = [2, 3]; b
@> let x = 3; let [_, b, (x)] = [1, 2, 3]; b

> @> let x = 3; let [b, (x)] = [2, 4]; b     # should fail

Reviewer: ```suggestion
@> let x = 3; let [b, (x)] = [2, 4]; b     # should fail because (x) != 4
```

> @> let [(x)] = [2]; x                      # should fail

Reviewer: ```suggestion
@> let [(x)] = [2]; x                      # should fail because `x` isn't in scope
```

```

`let a = 56; let {"x": a, "y": (a)} = {"x": 42, "y": 56}; a` is valid 
but `let a = 56; let {"x": a, "y": (a)} = {"x": 42, "y": 42}; a` should fail. 

> The previous statement could confuse people, should be avoided.

Reviewer: ```suggestion
Using the same name in an expression, `(a)`, and a newly bound name, `a`, is
confusing and should be avoided.
```

> What's more, Arr-ai allows extra elements `...` or `...x` in addition to the explicitly bounds ones and binds name `x` to the extra ones:

Reviewer: ```suggestion
What's more, arr.ai allows extra elements `...` or `...x` in addition to the explicitly bound ones and binds name `x` to the any additional elements that weren't explicitly matched by other patterns.
```
